dfp/decimal/decimal

```diff
@@ -209,38 +489,42 @@ namespace decimal
     }
 
 //  ISO/IEC TR 27433 - 3.2.11 Formatted input:
-
   template <class charT, class traits>
-    inline std::basic_istream<charT, traits> &
-      operator>>(std::basic_istream<charT, traits> & is, decimal32 & d)
-      {
-	char buf[CHAR_MAX];
-	memset(buf, '\0', CHAR_MAX);
-	is.read(buf,CHAR_MAX);
-	d.__setval(strtod32(buf, NULL));
-	return is;
-      }
+    inline std::dfp_basic_istream<charT, traits> &
+      operator>>(std::dfp_basic_istream<charT, traits> & is, decimal32 & d)
+      { return is._M_extract(d); }
 
   template <class charT, class traits>
-    inline std::basic_istream<charT, traits> &
-      operator>>(std::basic_istream<charT, traits> & is, decimal64 & d)
-      {
-	char buf[CHAR_MAX];
-	memset(buf, '\0', CHAR_MAX);
-	is.read(buf,CHAR_MAX);
-	d.__setval(strtod64(buf, NULL));
-	return is;
-      }
+    inline std::dfp_basic_istream<charT, traits> &
+      operator>>(std::dfp_basic_istream<charT, traits> & is, decimal64 & d)
+      { return is._M_extract(d); }
 
   template <class charT, class traits>
+    inline std::dfp_basic_istream<charT, traits> &
+      operator>>(std::dfp_basic_istream<charT, traits> & is, decimal128 & d)
+      { return is._M_extract(d); }
+
+  template <class charT, class traits, class T>
     inline std::basic_istream<charT, traits> &
-      operator>>(std::basic_istream<charT, traits> & is, decimal128 & d)
+      operator>>(std::basic_istream<charT, traits> & is, T & d)
       {
-	char buf[CHAR_MAX];
-	memset(buf, '\0', CHAR_MAX);
-	is.read(buf,CHAR_MAX);
-	d.__setval(strtod128(buf, NULL));
-	return is;
+        static_assert(
+            std::is_same<T, decimal::decimal32>::value
+              || std::is_same<T, decimal::decimal64>::value
+              || std::is_same<T, decimal::decimal128>::value,
+            "This template function is only meant to be used with "
+            "decimal floating point types"
+        );
+
+        auto& dfpIstream {
```

Comment:
Will this only work with C++11 and newer?

Reply:
`is_same` is trivial to reimplement in C++98.